### ISSUE TITLE
ESMF wants NetCDF and NetCDF-Fortran in the Same Directory

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#2809]
As of NetCDF 4.2, the C and Fortran libraries have been split into two different tarballs. Since the versions don't line up, we decided to split them into 2 packages, `netcdf` and `netcdf-fortran`. However, some build systems still expect them to be installed in the same directory. I'm currently packaging ESMF (the Earth System Modeling Framework), and their [user guide](http://www.earthsystemmodeling.org/esmf_releases/last_built/ESMF_usrdoc/node9.html) has this to say:

> To compile ESMF with NetCDF 4.2 and newer releases, it is required to have the NetCDF C library, and the NetCDF Fortran API library installed in the same directory.

This is easy to solve in the NetCDF-Fortran package. We would simply change:
```python
configure('--prefix={0}'.format(prefix))
```
to
```python
configure('--prefix={0}'.format(spec['netcdf'].prefix))
```
How do others feel about this? @citibeth?

**Comment by tgamblin** [comment]
I think that's the wrong way to do it.  If you do that, then packages that expect the NetCDF-Fortran interfaces to be in *its* prefix will break.

Do you know why they require these things to be in the same directory?  Is that just an artifact of their build system taking only one parameter?  Given that Spack injects RPATHs and includes, have you tried seeing if the build will work within Spack even with separate directories?

**Comment by davydden** [comment]
I would either join them in a single package or patch the build system of ESMF. If we hack installation of Fortran part outside of prefix, we would loose all versioning anyway (think of building two different Fortran parts against the same C). The suggested hack is also against the spack ideas

**Comment by citibeth** [comment]
We had a big discussion on this at the time.  Why not make a new package
netcdf-all that depends on netcdf and netcdf-fortran, then symlinks them
into spec.prefix?

**Comment by adamjstewart** [comment]
If I recall, the last time I spoke with the developers and asked if symlinks would work, they said it wouldn't.

**Comment by adamjstewart** [comment]
I'm not sure how easy it would be to patch the build system of ESMF.

**Comment by tgamblin** [comment]
The key thing we want to preserve with separate prefixes is that when two packages says they depends on `foo` or `foo+bar` etc., that the mean the same thing.  If installing `netcdf-fortran` modifies the prefix of `netcdf-c`, that changes the behavior of `netcdf-c` without reflecting the change in the `spec`, so other packages can't know whether *their* builds will conflict with the `netcdf-c` they depended on anymore.

If it were me the first thing I'd do would be to try to patch the build system, but I haven't looked at it.

I kind of prefer @citibeth's suggestion of symlinking both, if you can pull it off.  That ensures that `netcdf-c` and `netcdf-fortran` are concretized together.  I'm really curious why the developers think that symlinks won't work for them...  any chance you can find out?

**Comment by davydden** [comment]
I also like the idea of an auxiliary package to symlink the two if that's easier than patching.

**Comment by adamjstewart** [comment]
I'll poke around the build system and contact the developers and see what I can find. In the meantime, I'll put together a WIP PR that hacks `netcdf-fortran` until we can come up with a better solution.

**Comment by davydden** [comment]
If u'll be hacking, symlink the other way around: C into Fortran prefix after installation.

**Comment by davydden** [comment]
(This shall not break Spack in any way, but still is a hack)

**Comment by adamjstewart** [comment]
I don't think symlinks will work, but we'll find out!

**Comment by davydden** [comment]
Or try copying the C prefix into Fortran version, if symlinks don't work.

**Comment by tgamblin** [comment]
@adamjstewart: I would just try the build with the two in separate directories first.

I looked at the ESMF build, and the only thing that seems to cause this requirement is that they have ONE variable for where NetCDF libs are located (`ESMF_NETCDF`), and they use it to set library search paths (`-L$ESMF_NETCDF/lib`, etc.).  But Spack is already going to put `-L` flags on the compile line for dependencies, so it might not even need you to set that variable.

**Comment by citibeth** [comment]
@adamjstewart  I would try the following (in order):

1. Follow suggestion of @tgamblin and just try to build it anyway, see if it works with Spack.

2. Patch the ESMF build to handle two separate directories.  This is almost certainly no more than a minor modification.

3. Try symlinks

4. Try hardlinks

Where I work, there is a directive from high up to use ESMF; but it is so hard to build that nobody does.  But we still need to.  Therefore, building ESMF has been on my "to do someday" list.

It would be great if you get ESMF built with (2).  If so, I would work with the right people where I work to get your patches integrated upstream.  The story that we need this to build with modern tools is compelling, as is the promise of an easier way to build ESMF.  I really think that's the best solution, in the sense that it improves the overall community.

> I'll put together a WIP PR that hacks netcdf-fortran until we can come up with a better solution.

We had exactly that in the past.  For example, see:
    https://github.com/LLNL/spack/pull/405/files

It was eventually removed after a LONG discussion; see #466.  I would rather not resurrect this idea; the thread in #466 explains why.

**Comment by tgamblin** [comment]
I'll help with (2) if (1) doesn't work.  I agree with @citibeth -- it's hard to imagine this being a major change, or something we can't just work around.

**Comment by mathomp4** [comment]
As I work on getting ESMF working in spack for me (issues incoming), I wanted to add to this that later versions of ESMF now have `ESMF_NETCDF=split` as an option. Might help ESMF

**Comment by citibeth** [comment]
Great!  It will be simplest if we can just support later versions and be
done with their headache

**Comment by alalazo** [comment]
This has been solved in the ESMF package in #2831
